File: tedonactivityresult/src/main/java/com/gun0912/tedonactivityresult/TedOnActivityResult.java
package com.gun0912.tedonactivityresult;

import android.content.Context;
import android.content.Intent;
import com.gun0912.tedonactivityresult.listener.OnActivityResultListener;

public class TedOnActivityResult {

  public static Builder with(Context context) {
    return new Builder(context);
  }


  public static class Builder {

    private Context context;
    private OnActivityResultListener listener;
    private Intent intent;

    public Builder(Context context) {
      this.context = context;
    }

    public void startActivityForResult() {
      ProxyActivity.startActivityForResult(context, intent, listener);
    }

    public Builder setListener(OnActivityResultListener listener) {
      this.listener = listener;
      return this;
    }

    public Builder setIntent(Intent intent) {
      this.intent = intent;
      return this;
    }

  }
}

File: tedonactivityresult-coroutine/src/main/java/com/gun0912/tedonactivityresult/coroutine/TedOnActivityResult.kt
package com.gun0912.tedonactivityresult.coroutine

import android.content.Context
import android.content.Intent
import com.gun0912.tedonactivityresult.ProxyActivity
import com.gun0912.tedonactivityresult.listener.OnActivityResultListener
import com.gun0912.tedonactivityresult.model.ActivityResult
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

object TedOnActivityResult {

    fun with(context: Context): Builder = Builder(context)

    class Builder(private val context: Context) {

        suspend fun startActivityForResult(intent: Intent): ActivityResult =
            suspendCoroutine { continuation ->
                val listener = OnActivityResultListener { resultCode, data ->
                    continuation.resume(ActivityResult(resultCode, data))
                }
                ProxyActivity.startActivityForResult(context, intent, listener)
            }
    }
}
